lib/puma/cluster/worker.rb

```diff
@@ -122,49 +132,111 @@ def run
             debug_loaded_extensions "Loaded Extensions - worker 0:"
           end
 
-          stat_thread ||= Thread.new(@worker_write) do |io|
-            Puma.set_thread_name "stat pld"
-            base_payload = "#{PIPE_PING}#{Process.pid}"
-
-            while true
-              begin
-                b = server.backlog || 0
-                r = server.running || 0
-                t = server.pool_capacity || 0
-                m = server.max_threads || 0
-                rc = server.requests_count || 0
-                bt = server.busy_threads || 0
-                payload = %Q!#{base_payload}{ "backlog":#{b}, "running":#{r}, "pool_capacity":#{t}, "max_threads":#{m}, "requests_count":#{rc}, "busy_threads":#{bt} }\n!
-                io << payload
-              rescue IOError
-                Puma::Util.purge_interrupt_queue
-                break
-              end
-              sleep @options[:worker_check_interval]
-            end
-          end
+          make_sure_pinging(server)
+
           server_thread.join
         end
 
+        if @mold
+          set_proc_title(role: "mold")
+
+          Signal.trap("SIGTERM") do
+            @worker_write << "#{PIPE_EXTERNAL_TERM}#{Process.pid}\n" rescue nil
+            @fork_pipe.close
+          end
+
+          worker_pids = []
+          Signal.trap "SIGCHLD" do
+            wakeup! if worker_pids.reject! do |p|
+              Process.wait(p, Process::WNOHANG) rescue true
+            end
+          end
+
+          @config.run_hooks(:on_mold_promotion, index, @log_writer, @hook_data)
+
+          make_sure_pinging(server)
+          wakeup!
+
+          begin
+            while (idx = PipeProtocols::Fork.read_from(@fork_pipe))
```

Comment:
this returns nil if the pipe closes, which it does on TERM above

Reply:
Was mistaken here; pipe reads return nil if all _writers_ close, but if you close the reader directly (as we do since the writer's not present in workers) it raises EOFError. Fixed in the final changes branch.